Updated TSDB documentation with additional details (#5706)
elastic/integrations: docs/developer_tsdb_migration_guidelines.md

> 1. **Datastream having type `logs` can be excluded from TSDB migration.**

Reviewer: ```suggestion
1. **Datastream having type `logs` are excluded from TSDB migration.**
```

Reviewer: Saying can be gives the idea that it is optional, but it is not.

2. **Modify the `kibana.version` to 8.8.2 within the manifest.yml file of the package.**
   ```
   conditions:

>      kibana.version: "^8.8.0"

Reviewer 2: Should we make this 8.8.2

Author: `8.8.0` is what we used as the target in majority of our integrations . 

Many of us use `8.8.2` stack to avoid memory issues reported in 8.8.0. 

But, as a `kibana.version condition`, 8.8.0 should suffice.

Reviewer 2: The para is going to confuse the reader either we say 8.8.2  or 8.8.0 not both

Author: Oh sure, i noticed the confusion. Will correct it

Author: Updated as 8.8.0

    elasticsearch:
      index_mode: "time_series"
    ```

>     If your datastream has more number of dimension fields, you can modify this limit by modifying index.mapping.dimension_fields.limit value as below. Please note, this feature is available only from Elasticstack version 8.6.

Reviewer 3: ```suggestion
    If your datastream has more number of dimension fields, you can modify this limit by modifying index.mapping.dimension_fields.limit value as below. Please note, this feature is available only from Kibana version 8.6.
```

    From the context of integrations that are related to products that are deployed on-premise, there exist certain fields that are part of every package and they are potential candidates of becoming dimension fields

    * `host.name`
    * `service.address`
    * `agent.id`

>     * `container.id`

Reviewer: From[ this comment](https://github.com/elastic/integrations/issues/5193#issuecomment-1536090359), only kubernetes and docker have value for `container.id`, but this part of the file gives the idea that `container.id` should always be a dimension. Can we restructure this to include the fields we consider "mandatory" like `agent.id` and why is that?

Author: For non-cloud packages, which can run as containers, container-id is a dimension. 
But, due to the limitations in elastic-agent this is not presently supported. This is the reason why in the comment "For now" clause is added. 

The recommendations based on the deployment environment is already part of this document.

    For products operating as managed services within cloud providers like AWS, Azure, and GCP, it is advised to label the fields listed below as dimension fields.
    * `cloud.account.id`
    * `cloud.region`
    * `cloud.availability_zone`
    * `cloud.provider`
    * `agent.id ` 
    

4. **Annotating the integration specific fields as dimension**

    `files.yml` file has the field mappings specific to a datastream of an integration. This step is needed when the dimension fields in ECS is not sufficient enough to create a unique [_tsid](https://www.elastic.co/guide/en/elasticsearch/reference/current/tsds.html#tsid) value for the documents stored in elasticsearch. Annotate the field with `dimension: true` to tag the field as dimension field. 

    Adding an inline comment prior to the dimension annotation is advised, detailing the rationale behind the choice of a particular field as a dimension field.

    ```
    - name: wait_class
      type: keyword
      # Multiple events are generated based on the values of wait_class. Hence, it is a dimension
      dimension: true
      description: Every wait event belongs to a class of wait events.
    ```
    *Notes:*

>     * *There exists a limit on how many dimension fields can have. By default this value is 21.*

Reviewer: Can you link to the [source](https://github.com/elastic/elasticsearch/blob/6417a4f80f32ace48b8ad682ad46b19b57e49d60/server/src/main/java/org/elasticsearch/index/mapper/MapperService.java#L114) of this? So it gets easier to fix this part in the future in case it changes.

Author: I think, changing the value from 21 to something higher is a very rare scenario. But changing the structure of the code especially the line number is highly possible, making the link point to a location that is not meaningful.

Reviewer: True, that is why we should use a permalink.

    * *Dimension keys have a hard limit of 512b. Documents are rejected if this limit is reached.*
    * *Dimension values have a hard limit of 1024b. Documents are rejected if this limit is reached*  

>     **Warning:** Choosing an insufficient number of dimension fields may lead to data loss

Reviewer: Can you refer the reader to the testing section and link the TSDB migration test kit?

Author: @constanca-m , can we take that as a separate update to this document? May be you can add that part?

 
- After migration, verify if the dashboard is rendering the data properly. If certain visualisation do not work, consider migrating to [Lens](https://www.elastic.co/guide/en/kibana/current/lens.html)

>   Certain aggregation functions are not supported when a field is having a metric_type `counter`. Example `avg()`. Replace such aggregation functions with a supported aggregation type such as `max()` or `min()`. 

Reviewer: `min` for counter doesn't seem to make sense, as that would always be the first value. I am also not sure that aggregation is supported, I think only avg and max, but I am not sure of this either. In any case, `min` does not seem to be a good reference on this part of the file.

Author: In the absence of avg() function, wouldn't the formula making use of min() and  max() within a specific period of time, help to determine the avg() ?

Reviewer: I checked and `min` is possible. But why would you try to compute average in a counter field?

Author: I am merely giving avg() as an example, making use of max() and min() value, if there exist a usecase.

- Currently, there are several limits around the number of dimensions.  
 Reference : https://github.com/elastic/elasticsearch/issues/93564

- Other known issues: https://github.com/elastic/integrations/issues/5233. Refer the section - New Issues Identified, TSDB Issues reported earlier.

# <a id="existing-migrated-packages"></a> Reference to existing package already migrated

> - Oracle integration TSDB enablement: [PR Link](https://github.com/elastic/integrations/pull/5307)

Reviewer 2: Let's point to the git repo instead.
